![](https://github.com/sbme-tutorials/Biostatistics-Tutorials/blob/main/Notebooks/images/Python_logo_and_wordmark.svg.png?raw=1)

<h2> General Info </h2>

* Interpreted language 
* Untyped language
* Cross platform
* Popular for 
    * Web Development.
    * Machine Learning and Artificial Intelligence.
    * Data Science and Data Visualization.
    * Desktop GUI
    * Game Development.
    * and more
   

<br>
<h2> Python Popular Frameworks and Libraries </h2>


Python has wide range of libraries and frameworks widely used in various fields such as machine learning, artificial intelligence, web applications, etc. We define some popular frameworks and libraries of Python as follows.

- Web development (Server-side) - Django Flask, Pyramid, CherryPy
- GUIs based applications - Tk, PyGTK, PyQt, PyJs, etc.
- Machine Learning - TensorFlow, PyTorch, Scikit-learn, Matplotlib, Scipy, etc.
- Mathematics - Numpy, Pandas, etc.

# Variables 

- In Python, we don't need to specify the type of variable because Python is a infer language and smart enough to get variable type.

<h3> variables type</h3>

1. local: Local variables are the variables that declared inside the function and have scope within the function.

2. Global: Global variables can be used throughout the program, and its scope is in the entire program. We can use global variables inside or outside the function.A variable declared outside the function is the global variable by default. Python provides the global keyword to use global variable inside the function. If we don't use the global keyword, the function treats it as a local variable. ` global x `


In [68]:
# Declaring a function  
def add():  
    # Defining local variables. They has scope only within a function  
    a = 20  
    b = 30  
    c = a + b  
    print("The sum is:", c)  
  
# Calling a function  
add()  

The sum is: 50


In [69]:
# Declare a variable and initialize it  
x = 101  
  
# Global variable in function  
def mainFunction():  
    # printing a global variable  
    global x  
    print(x)  
    # modifying a global variable  
    x = 'Welcome To Javatpoint'  
    print(x)  
  
mainFunction()  
print(x)  

101
Welcome To Javatpoint
Welcome To Javatpoint


<h3> Delete Variable </h3>

- We can delete the variable using the del keyword 


In [70]:
# Assigning a value to x  
x = 6  
print(x)  
# deleting a variable.

del x  




6


<h2> maximam size of int or float</h2>

- there is no limitation number by bits and we can expand to the limit of our memory.

- Python doesn't have any special data type to store larger numbers.

<h2> Printing Multiple Variables</h2>


In [71]:
a = 5  
b = 6  
# printing multiple variables  
print(a,b)  
# separate the variables by the comma  
print (1, 2, 3, 4, 5, 6, 7, 8)   

5 6
1 2 3 4 5 6 7 8


# Data Types 

- Variables can hold values, and every value has a data-type. Python is a dynamically typed language; hence we do not need to define the type of the variable while declaring it. The interpreter implicitly binds the value with its type

In [72]:
a=10  
b="Hi Python"  
c = 10.5  
print(type(a))  
print(type(b))  
print(type(c))  

<class 'int'>
<class 'str'>
<class 'float'>


<h3> Standard data types </h3>

- A variable can hold different types of values. For example, a person's name must be stored as a string whereas its id must be stored as an integer.

- Python provides various standard data types that define the storage method on each of them. The data types defined in Python are given below.



<img src =" https://static.javatpoint.com/python/images/python-data-types.png ">

<br><br>

<h3> 

1. Numeric
</h3>

In [73]:
# 
a = 5  
print("The type of a", type(a))  
  
b = 40.5  
print("The type of b", type(b))  
  
c = 1+3j  
print("The type of c", type(c))  
print(" c is a complex number", isinstance(1+3j,complex))  

The type of a <class 'int'>
The type of b <class 'float'>
The type of c <class 'complex'>
 c is a complex number True


<h3> 

2. Sequence Type
</h3>


<h4> String </h4>
<br>
<p>

* The string can be defined as the sequence of characters represented in the quotation marks. In Python, we can use single, double, or triple quotes to define a string.

* String handling in Python is a straightforward task since Python provides built-in functions and operators to perform operations in the string.

* In the case of string handling, the operator + is used to concatenate two strings as the operation "hello"+" python" returns "hello python".

* The operator * is known as a repetition operator as the operation "Python" *2 returns 'Python Python'.



</p>

In [74]:
str1 = 'hello javatpoint' #string str1    
str2 = ' how are you' #string str2    
print (str1[0:2]) #printing first two character using slice operator    
print (str1[4]) #printing 4th character of the string    
print (str1*2) #printing the string twice    
print (str1 + str2) #printing the concatenation of str1 and str2    

he
o
hello javatpointhello javatpoint
hello javatpoint how are you


<h4> types of string </h4>

1. Single-line String- Strings that are terminated within a single-line are known as Single line Strings.
2. Multi-line String - A piece of text that is written in multiple lines is known as multiple lines string


In [75]:
# there are tow way to write multi-line string 
# 1 Adding black slash at the end of each line.
text1="osamah faisal\
 abdlatef"    
print(text1)  

# 2  Using triple quotation marks
text2=""" osamah Faisal
Abdulatef """
print(text2)


osamah faisal abdlatef
 osamah Faisal
Abdulatef 


<h4> string Formating </h4>

* as the following we can add int or string or any thing to string 

In [76]:
# This prints out "Hello, John!"
name = "John"
print("Hello, %s!" % name)

# This prints out "John is 23 years old."
name = "John"
age = 23
print("%s is %d years old." % (name, age))

# This prints out: A list: [1, 2, 3]
mylist = [1,2,3]
print("A list: %s" % mylist)


data = ("John", "Doe", 53.44) # tuple
# Hello John Doe. Your current balance is $53.44.
string_to_print = "Hello %s %s. Your current balance is $%s"
print(string_to_print%data)

Hello, John!
John is 23 years old.
A list: [1, 2, 3]
Hello John Doe. Your current balance is $53.44


<h4> List </h4>

* Python Lists are similar to arrays in C. However, the list can contain data of different types. The items stored in the list are separated with a comma (,) and enclosed within square brackets [].

* We can use slice [:] operators to access the data of the list. The concatenation operator (+) and repetition operator (*) works with the list in the same way as they were working with the strings.
* we can add element in the end or pop it by this way list.appaned(5), list.pop()

In [77]:
list1  = [1, "hi", "Python", 2]    
#Checking type of given list  
print(type(list1))  
  
#Printing the list1  
print (list1)  
  
# List slicing  
print (list1[3:])  
  
# List slicing  
print (list1[0:2])   
  
# List Concatenation using + operator  
print (list1 + list1)  
  
# List repetation using * operator  
print (list1 * 3)  

# slice by step as print(list[start : end : step ])
print(list1[0:3:2])

# slice by step from begin to  end 
print (list1[: :2])

# to print list from end 
print(list1[::-1])

<class 'list'>
[1, 'hi', 'Python', 2]
[2]
[1, 'hi']
[1, 'hi', 'Python', 2, 1, 'hi', 'Python', 2]
[1, 'hi', 'Python', 2, 1, 'hi', 'Python', 2, 1, 'hi', 'Python', 2]
[1, 'Python']
[1, 'Python']
[2, 'Python', 'hi', 1]


<h4> built in function in list </h4>

1. length of the list      ` len(list)`
2. max value in the list     ` max(list)`
3. min value in the list    ` min(list) `
4. convert any squence to  list     ` list(anySqu) `

<h4> Tuple </h4>

* A tuple is similar to the list in many ways. Like lists, tuples also contain the collection of the items of different data types. The items of the tuple are separated with a comma (,) and enclosed in parentheses ().

* A tuple is a read-only data structure as we can't modify the size and value of the items of a tuple.

* Let's see a simple example of the tuple.

In [78]:
tup  = ("hi", "Python", 2)    
# Checking type of tup  
print (type(tup))    
  
#Printing the tuple  
print (tup)  
  
# Tuple slicing  
print (tup[1:])    
print (tup[0:1])    
  
# Tuple concatenation using + operator  
print (tup + tup)    
  
# Tuple repatation using * operator  
print (tup * 3)     
  
# Adding value to tup. It will throw an error.  
# tup[2] = "hi"  

<class 'tuple'>
('hi', 'Python', 2)
('Python', 2)
('hi',)
('hi', 'Python', 2, 'hi', 'Python', 2)
('hi', 'Python', 2, 'hi', 'Python', 2, 'hi', 'Python', 2)


<h4> Dictionary </h4>

* Dictionary is an unordered set of a key-value pair of items. It is like an associative array or a hash table where each key stores a specific value. Key can hold any primitive data type, whereas value is an arbitrary Python object.

* The items in the dictionary are separated with the comma (,) and enclosed in the curly braces {}.

* Consider the following example.

In [79]:
d = {1:'Jimmy', 2:'Alex', 3:'john', 4:'mike'}     
  
# Printing dictionary  
print (d) 

#add value
d[5]='saad' 
  
# Accesing value using keys  
print("1st name is "+d[1])   
print("2nd name is "+ d[4])    
  
print (d.keys())    
print (d.values())  
 
# we can do update or delete 
d.update({1:'Fasil' , 2:'abduslam'})
print (d.values()) 

del d[1]
print (d.values())

# copy the dictionery 
d2=d.copy()

# clear the dictionery 
d.clear() 

print("this is capy of dic %s" , list(d2))

{1: 'Jimmy', 2: 'Alex', 3: 'john', 4: 'mike'}
1st name is Jimmy
2nd name is mike
dict_keys([1, 2, 3, 4, 5])
dict_values(['Jimmy', 'Alex', 'john', 'mike', 'saad'])
dict_values(['Fasil', 'abduslam', 'john', 'mike', 'saad'])
dict_values(['abduslam', 'john', 'mike', 'saad'])
this is capy of dic %s [2, 3, 4, 5]


<h4> set </h4>

* Python Set is the unordered collection of the data type. It is iterable, mutable(can modify after creation), and has unique elements. In set, the order of the elements is undefined; it may return the changed sequence of the element. The set is created by using a built-in function set(), or a sequence of elements is passed in the curly braces and separated by the comma. It can contain various types of values. Consider the following example.

* Unlike other collections in Python, there is no index attached to the elements of the set, i.e., we cannot directly access any element of the set by the index. However, we can print them all together, or we can get the list of elements by looping through the set


In [80]:
Days = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"}    
print(Days)    
print(type(Days))    
print("looping through the set elements ... ")    
for i in Days:    
    print(i)    

{'Sunday', 'Friday', 'Wednesday', 'Tuesday', 'Saturday', 'Monday', 'Thursday'}
<class 'set'>
looping through the set elements ... 
Sunday
Friday
Wednesday
Tuesday
Saturday
Monday
Thursday


# keyword 

* Python keywords are unique words reserved with defined meanings and functions that we can only apply for those functions


<img src =" Screenshot 2022-08-15 022427.jpg ">


* to  show keyword in your version we can do that in the ide as the flowing 
<br>
`import keyword  ` 
<br>
`print(keyword.kwlist) `

<br>

<img src="Screenshot 2022-08-15 023726.jpg">





<h4> in keyword</h4>

* The in keyword of Python is a robust confinement checker, also known as a membership operator. If you provide it an element to seek and a container or series to seek into, it will give True or False, depending on if that given element was located in the given container:

In [81]:
str1= "osamah "
print('o' in str1)
print('1' in str1)

True
False


<h4> is Keyword <h4>

* When the objects are the same, it gives True; otherwise, it gives False.

* A blank dictionary or list is the same as another blank one. However, they aren't identical entities because they are stored independently in memory. This is because both the list and the dictionary are changeable.

In [82]:
print( True is True )  
print( False is True )  
print( None is not None )  
print( (9 + 5) is (7 * 2) )  

True
False
False
True


<>:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\hp\AppData\Local\Temp\ipykernel_4492\2135147946.py:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
  print( (9 + 5) is (7 * 2) )


<h4> Iteration Keywords: for, while, break, continue </h4>

* The iterative process and looping are essential programming fundamentals. To generate and operate with loops, Python has multiple keywords. These would be utilized and observed in almost every Python program. Knowing how to use them correctly can assist you in becoming a better Python developer.

1. The for Keyword

The for loop is by far the most popular loop in Python. It's built by blending two Python keywords. They are for and in, as previously explained.

2. The while Keyword

Python's while loop employs the term while and functions similarly to other computer languages' while loops. The block after the while phrase will be repeated repeatedly until the condition following the while keyword is false.

3. The break Keyword

If you want to quickly break out of a loop, employ the break keyword. We can use this keyword in both for and while loops.

4. The continue Keyword

You can use the continue Python keyword if you wish to jump to the subsequent loop iteration. The continue keyword, as in many other computer languages, enables you to quit performing the present loop iteration and go on to the subsequent one

In [83]:
# Program to show the use of keywords for, while, break, continue  
for i in range(15):  
    
    print( i + 4, end = " ")  
        
    # breaking the loop when i = 9  
    if i == 9:  
        break     
print()  
        
# looping from 1 to 15  
i = 0 # initial condition  
while i < 15:  
        
    # When i has value 9, loop will jump to next iteration using continue. It will not print  
    if i == 9:  
        i += 3  
        continue  
    else:  
        # when i is not equal to 9, adding 2 and printing the value  
        print( i + 2, end = " ")  
            
    i += 1  

4 5 6 7 8 9 10 11 12 13 
2 3 4 5 6 7 8 9 10 14 15 16 

# function 

*  Function Arguments

The following are the types of arguments that we can use to call a function:

1. Default Arguments

A default argument is a kind of parameter that takes as input a default value if no value is supplied for the argument when the function is called.

In [84]:
def func1(x1 , x2=30):
    return (x1+x2)

print(func1(1))
print(func1(1 , 4))

31
5


2. keyword arguments

The arguments in a function called are connected to keyword arguments. If we provide keyword arguments while calling a function, the user uses the parameter label to identify which parameters value it is.


In [85]:

def func2(num1 , num2):
    print("num 1 is " , num1)
    print("num 2 is " , num2)

func2(num2=75 , num1=30)

num 1 is  30
num 2 is  75


3. required arguments

the argument in anu function not has any value so most pass value when calling function 

4. Variable-Length Arguments

We can use special characters in Python functions to pass as many arguments as we want in a function. There are two types of characters that we can use for this purpose:

* *args -These are Non-Keyword Arguments

* **kwargs - These are Keyword Arguments.

In [86]:
# *args
def func3(*args):
    list=[]
    for i in args:
        list.append(i)
    print(list)
    
object=(1 , 2 , 3 , 4 , 4 , 5 , 6 )

func3(object)

[(1, 2, 3, 4, 4, 5, 6)]


# range Function 

* With the help of the range() function, we may produce a series of numbers. range(10) will produce values between 0 and 9. (10 numbers).

* We can give specific start, stop, and step size values in the manner range(start, stop, step size). If the step size is not specified, it defaults to 1.

* Since it doesn't create every value it "contains" after we construct it, the range object can be characterized as being "slow." It does provide in, len, and __getitem__ actions, but it is not an iterator.

In [87]:
# Python program to iterate over a sequence with the help of indexing  
  
tuple_ = ("Python", "Loops", "Sequence", "Condition", "Range")  
  
# iterating over tuple_ using range() function  
for iterator in range(len(tuple_)):  
    print(tuple_[iterator].upper())  

PYTHON
LOOPS
SEQUENCE
CONDITION
RANGE


# File I/O

* Sometimes, it is not enough to only display the data on the console. The data to be displayed may be very large, and only a limited amount of data can be displayed on the console since the memory is volatile, it is impossible to recover the programmatically generated data again and again.

* The file handling plays an important role when the data needs to be stored permanently into the file. A file is a named location on disk to store related information. We can access the stored information (non-volatile) after the program termination.

* In Python, files are treated in two modes as text or binary. The file may be in the text or binary format, and each line of a file is ended with the special character.

* a file operation can be done in the following order.
1. open file 
2. read or write 
3. close file 


1. openning file  and read or write 

* Python provides an open() function that accepts two arguments, file name and access mode in which the file is accessed. The function returns a file object which can be used to perform various operations like reading, writing, etc

* syntax of open file `  fileObject=open("fileDirector" , "operation -as rea-")`

* the following table show the operation can do it in the open file 

<img src ="Screenshot 2022-08-16 065056.jpg">



<h4> Read file </h4>

* To read a file using the Python script, the Python provides the read() method. The read() method reads a string from the file. It can read the data in the text as well as a binary format.

* the syntax of read in  --> `fileObject.read( number of conant) `

<h4> For loop to reading file </h4>

`#open the file.txt in read mode. causes an error if no such file exists. `

` fileptr = open("file2.txt","r"); `    

` #running a for loop`     

`for i in fileptr:    `

`print(i) # i contains each line of the file`




<h4> Read Lines of the file </h4>

* Python facilitates to read the file line by line by using a function readline() method. The readline() method reads the lines of the file from the beginning, i.e., if we use the readline() method two times, then we can get the first two lines of the file. if we want to read all  line we use readlines()



<h4> Writing the file </h4>

* To write some text to a file, we need to open the file using the open method with one of the following access modes.

* w: It will overwrite the file if any file exists. The file pointer is at the beginning of the file.

* a: It will append the existing file. The file pointer is at the end of the file. It creates a new file if no file exists.


In [88]:
# open file with read only 
FileRead=open("FileToTest.tex" , "r")
print(FileRead.read(100))

# file to write  
FileWrite=open("FileWrite.tex" ,"w")
FileWrite.write("osamah Faisal Abdulatef")

# file append 

FileAppend=open("FileToTest.tex" , "a")
FileAppend.write(" OSamah Faisal")




osamah Faial Abdulatef  OSamah Faisal OSamah Faisal OSamah Faisal OSamah Faisal


14

2. close file 

* Once all the operations are done on the file, we must close it through our Python script using the close() method. Any unwritten information gets destroyed once the close() method is called on a file object.

* We can perform any operation on the file externally using the file system which is the currently opened in Python; hence it is good practice to close the file once all the operations are done.

* syntax of vlose file  -->  ` fileObject.close()`

* After closing the file, we cannot perform any operation in the file. The file needs to be properly closed. If any exception occurs while performing some operations in the file then the program terminates without closing the file

In [89]:
# we can read and print data in the consl by the way

file1=open("FileToTest.tex" , "r+")
file1.write("osamah Faial Abdulatef")
# becouse in this way the data don't save in the file so close and open the file to write data 
file1.close()

file1=open("FileToTest.tex" , "r")
contant=file1.read()
print(type(contant))
print(contant)
file1.close()


<class 'str'>
osamah Faial Abdulatef  OSamah Faisal OSamah Faisal OSamah Faisal OSamah Faisal OSamah Faisal


<h4> The with statement </h4>

* The with statement was introduced in python 2.5. The with statement is useful in the case of manipulating the files. It is used in the scenario where a pair of statements is to be executed with a block of code in between.

* The advantage of using with statement is that it provides the guarantee to close the file regardless of how the nested block exits.

* It is always suggestible to use the with statement in the case of files because, if the break, return, or exception occurs in the nested block of code then it automatically closes the file, we don't need to write the close() function. It doesn't let the file to corrupt.

In [90]:
with open("FileToTest.tex" ,"r") as file:
    contant=file.read()
    print(contant)

file.close()


osamah Faial Abdulatef  OSamah Faisal OSamah Faisal OSamah Faisal OSamah Faisal OSamah Faisal


<h4> File pointer position</h4> 

* Python provides the tell() method which is used to print the byte number at which the file pointer currently exists.

In [91]:
 
#  pointer befor reading and after reading 

file=open("FiletoTest.tex", "r")
# befor reading 
print(file.tell())

contant=file.read()
# after reading the file 
print(file.tell())


0
93


<h4> Modifying file pointer position </h4>

* In real-world applications, sometimes we need to change the file pointer location externally since we may need to read or write the content at various locations.

* For this purpose, the Python provides us the seek() method which enables us to modify the file pointer position externally

* **The seek() method accepts two parameters:**

1. offset: It refers to the new position of the file pointer within the file.

2. from: It indicates the reference position from where the bytes are to be moved. If it is set to 0, the beginning of the file is used as the reference position. If it is set to 1, the current position of the file pointer is used as the reference position. If it is set to 2, the end of the file pointer is used as the reference position.

In [92]:
file2=open("FileToTest.tex" , "r")
#befor seeking 
print(file2.tell())


file2.seek(10)
# after seeking
print(file2.tell())


0
10


<h4> Python OS module </h4>

* The Python os module enables interaction with the operating system. The os module provides the functions that are involved in file processing operations like renaming, deleting, etc

1. Rename

* first import os modul 
* syntax --> ` os.rename("lastname" , " newname") `

2. remove file 

* syntax --> ` os.remove("filename")`

3. create new directory 

* The mkdir() method is used to create the directories in the current working directory. 

`import os `

`os.mkdir("new")  `

4. The getcwd() method

* This method returns the current working directory.


5. Changing the current working directory

* The chdir() method is used to change the current working directory to a specified directory.

6. Deleting directory

* The rmdir() method is used to delete the specified directory.


<h4> Writing Python output to the files </h4>


** important not study ** 

* In Python, there are the requirements to write the output of a Python script to a file.

* The check_call() method of module subprocess is used to execute a Python script and write the output of that script to a file.

* The following example contains two python scripts. The script file1.py executes the script file.py and writes its output to the text file output.txt.

<h4> summery of the file function </h4>


<img src ="Screenshot 2022-08-16 091339.jpg" >

# python modules 

* how to access module in the python:- there are many way

1. ` import name_of_module `
2. `import name_of_module as rename_module ` 
3. ` from namae_of_module import as rename_module` 
4. import all  library from module as --> ` from name_of_module import *` 
 

# exception 

* An exception in Python is an incident that happens while executing a program that causes the regular course of the program's commands to be disrupted. When a Python code comes across a condition it can't handle, it raises an exception. An object in Python that describes an error is called an exception.

* When a Python code throws an exception, it has two options: handle the exception immediately or stop and quit.

<h4> Try and Except Statement - Catching Exceptions</h4>

* In Python, we catch exceptions and handle them using try and except code blocks. The try clause contains the code that can raise an exception, while the except clause contains the code lines that handle the exception.


In [93]:
# Python code to catch an exception and handle it using try and except code blocks  
   
a = ["Python", "Exceptions", "try and except"]  
try:  
    #looping through the elements of the array a, choosing a range that goes beyond the length of the array  
     for i in range( 4 ):  
        print( "The index and element from the array is", i, a[i] )  
#if an error occurs in the try block, then except block will be executed by the Python interpreter       
except:  
    print ("Index out of range")  

The index and element from the array is 0 Python
The index and element from the array is 1 Exceptions
The index and element from the array is 2 try and except
Index out of range


In [94]:
# x="true"
# while (x=="true"):
#     try: 
#         ID = int(input())
#         x="false"
       
#     except:
#         print("please Enter Your ID Again")
    
# print("done")

<h4> How to Raise an Exception </h4>

* If a condition does not meet our criteria but is correct according to the Python interpreter, we can intentionally raise an exception using the raise keyword.

* if the condition accur the program stop and show the exception you raise it .

In [95]:
# #Python code to show how to raise an exception in Python  
# num = [3, 4, 5, 7]  
# if len(num) > 3:  
#     raise Exception( f"Length of the given list must be less than or equal to 3 but is {len(num)}" ) 

In [96]:
# from logging import exception


# ID=int(input())
# if ID <= 100000:
#             raise exception(f" the number Entered is {ID} please try again")
        

<h4> Finally Keyword in Python </h4>

* The finally keyword is available in Python, and it is always used after the try-except block. The finally code block is always executed after the try block has terminated normally or after the try block has terminated for some other reason.

<h4> Summary </h4>

* We can throw an exception at any line of code using the raise keyword.
* Using the assert keyword, we may check to see if a specific condition is fulfilled and raise an exception if it is not.
* All statements are carried out in the try clause until an exception is found.
* The try clause's exception(s) are detected and handled using the except function.
* If no exceptions are thrown in the try code block, we can write code to be executed in the else code block.

In [97]:
''' # summery of exception 
try:  
    # Code block  
    # These statements are those which can probably have some error  
  
except:  
    # This block is optional.  
    # If the try block encounters an exception, this block will handle it.  
  
else:  
    # If there is no exception, this code block will be executed by the Python interpreter  
  
finally:  
    # Python interpreter will always execute this code. 

'''
 

' # summery of exception \ntry:  \n    # Code block  \n    # These statements are those which can probably have some error  \n  \nexcept:  \n    # This block is optional.  \n    # If the try block encounters an exception, this block will handle it.  \n  \nelse:  \n    # If there is no exception, this code block will be executed by the Python interpreter  \n  \nfinally:  \n    # Python interpreter will always execute this code. \n\n'

#  Date and time

* Python provides the datetime module work with real dates and times. In real-world applications, we need to work with the date and time. Python enables us to schedule our Python script to run at a particular timing.

* In Python, the date is not a data type, but we can work with the date objects by importing the module named with datetime, time, and calendar.

<h4> datetime </h4>

The datetime classes are classified in the six main classes.

* date - It is a naive ideal date. It consists of the year, month, and day as attributes.
* time - It is a perfect time, assuming every day has precisely 24*60*60 seconds. It has hour, minute, second, microsecond, and tzinfo as attributes.
* datetime - It is a grouping of date and time, along with the attributes year, month, day, hour, minute, second, microsecond, and tzinfo.
* timedelta - It represents the difference between two dates, time or datetime instances to microsecond resolution.
* tzinfo - It provides time zone information objects.
* timezone - It is included in the new version of Python. It is the class that implements the tzinfo abstract base class.

In [98]:
# to get current time 
import time;    
    
#returns a time tuple     
    
print(time.localtime())  

time.struct_time(tm_year=2022, tm_mon=8, tm_mday=17, tm_hour=10, tm_min=30, tm_sec=23, tm_wday=2, tm_yday=229, tm_isdst=0)


<h4> Time tuple </h4>

* the time is treated as the tuple of 9 numbers. Let's look at the members of the time tuple. 

* time.struct_time(tm_year=2022, tm_mon=8, tm_mday=17, tm_hour=10, tm_min=30, tm_sec=23, tm_wday=2, tm_yday=229, tm_isdst=0)



<img src="Screenshot 2022-08-17 103158.jpg">

<h4>sleep time </h4>

* The sleep() method of time module is used to stop the execution of the script for a given amount of time. The output will be delayed for the number of seconds provided as the float

In [99]:
import time  
for i in range(0,5):  
    print(i)  
    #Each element will be printed after 1 second  
    time.sleep(1)

0
1
2
3
4


<h4> Gitting time formate</h4>

* The time can be formatted by using the asctime() function of the time module. 


In [101]:
#  time formate 1
import time    
  #returns the formatted time        
print(time.asctime()) 



# from tuple time we can access any element 
import time 
t=time.localtime()
print(t[0])

Wed Aug 17 10:51:20 2022
2022


<h4> datetime module </h4>

* The datetime module enables us to create the custom date objects, perform various operations on dates like the comparison, etc.



In [103]:
import datetime  
#returns the current datetime object     
print(datetime.datetime.now()) 

2022-08-17 10:54:13.396488


<h4> Creating date objects </h4>

* We can create the date objects bypassing the desired date in the datetime constructor for which the date objects are to be created.

In [104]:
import datetime  
    
#returns the datetime object for the specified time      
    
print(datetime.datetime(2020,4,4,1,26,40))    

2020-04-04 01:26:40


In [108]:
import datetime
print(datetime.datetime(2020,12,10))

2020-12-10 00:00:00


<h4> Comparison of two dates </h4>

* We can compare two dates by using the comparison operators like >, >=, <, and <=.

In [110]:
from datetime import datetime as dt    
#Compares the time. If the time is in between 8AM and 4PM, then it prints working hours otherwise it prints fun hours    
if dt(dt.now().year,dt.now().month,dt.now().day,8)<dt.now()<dt(dt.now().year,dt.now().month,dt.now().day,16):    
    print("Working hours....")    
else:    
    print("fun hours")   

Working hours....


#  read csv file

A csv stands for "comma separated values", which is defined as a simple file format that uses specific structuring to arrange tabular data. It stores tabular data such as spreadsheet or database in plain text and has a common format for data interchange.


 <h4> csv Module Functions </h4>

The CSV module work is used to handle the CSV files to read/write and get data from specified columns. There are different types of CSV functions, which are as follows:

* csv.field_size_limit - It returns the current maximum field size allowed by the parser.
* csv.get_dialect - It returns the dialect associated with a name.
* csv.list_dialects - It returns the names of all registered dialects.
* csv.reader - It read the data from a csv file
* csv.register_dialect - It associates dialect with a name. The name must be a string or a Unicode object.
* csv.writer - It writes the data to a csv file
* o csv.unregister_dialect - It deletes the dialect which is associated with the name from the dialect registry. If a name is not a registered dialect name, then an error is being raised.
* csv.QUOTE_ALL - It instructs the writer objects to quote all fields. csv.QUOTE_MINIMAL - It instructs the writer objects to quote only those fields which contain special characters such as quotechar, delimiter, etc.
* csv.QUOTE_NONNUMERIC - It instructs the writer objects to quote all the non-numeric fields.
* csv.QUOTE_NONE - It instructs the writer object never to quote the fields.

<h4> Reading CSV files with csv lib</h4>

* Python provides various functions to read csv file. We are describing few method of reading function.

* Using csv.reader() function
* In Python, the csv.reader() module is used to read the csv file. It takes each row of the file and makes a list of all the columns.

In [ ]:
import csv  
  
# with open('csv_file.csv') as csv_file: 
csv_file=open('csv_file.csv')

#  some time a delimiter not eqaul , so when not eqaul  using following form 
csv_reader = csv.reader(csv_file, delimiter=',')    # this form when delimeter not equal  ,  
line_count = 0    
for i in csv_reader:    
    if line_count == 0:    
        print(f'Column names are {i}')    
        line_count += 1 
    else:
        print(f'row data are{i}')

csv_file.close()

<h4> Reading csv files with Pandas </h4>

* The Pandas is defined as an open-source library which is built on the top of the NumPy library. It provides fast analysis, data cleaning, and preparation of the data for the user.

* Reading the csv file into a pandas DataFrame is quick and straight forward. We don't need to write enough lines of code to open, analyze, and read the csv file in pandas and it stores the data in DataFrame.



In [4]:
# 1
from operator import index
import pandas

my_data=pandas.read_csv("pandas_file.txt" , delimiter=",") # create file my_data and put data of file pandas_file.txt within it 
print(my_data ) # print file my_data

my_data.to_csv("pandas_file.csv" ) # convert file my_data from txt to csv 
# my_data.to_csv("pandas_file.csv"  , index=None ) # index=None that do to cancel index in the axel file 
# my_data.to_csv("pandas_file.tex" , index=None )



# 2
# import pandas
# my_data=open("pandas_file.txt")
# print(pandas.read_csv(my_data))

   Unnamed: 0            Name Hire Date   Salary  Leaves Remaining    
0           0       John Idle  08/15/14  50000.0                    10
1           1   Smith Gilliam  04/07/15  65000.0                     8
2           2  Parker Chapman  02/21/14  45000.0                    10
3           3     Jones Palin  10/14/13  70000.0                     3
4           4   Terry Gilliam  07/22/14  48000.0                     7
5           5   Michael Palin  06/28/13  66000.0                     8


# Python OOPs Concepts

Major principles of object-oriented programming system are given below.

- Class
- Object
- Method
- Inheritance
- Polymorphism
- Data Abstraction
- Encapsulation

<h4> Class </h4>
The class can be defined as a collection of objects. It is a logical entity that has some specific attributes and methods.

- syntax 

```
class className:

    <statement-1>     
        .     
        .      
    <statement-N>   

```


<h4> Object </h4>


- The object is an entity that has state and behavior. It may be any real-world object like the mouse, keyboard, chair, table, pen, etc.

- Everything in Python is an object, and almost everything has attributes and methods. All functions have a built-in attribute __doc__, which returns the docstring defined in the function source code.

In [2]:
class car:  
    def __init__(self,modelname, year):  
        self.modelname = modelname  
        self.year = year  
    def display(self):  
        print(self.modelname,self.year)  
  
c1 = car("Toyota", 2016)  
c1.display()  

Toyota 2016


- In the above example, we have created the class named car, and it has two attributes modelname and year. We have created a c1 object to access the class attribute. The c1 object will allocate memory for these values. We will learn more about class and object in the next tutorial.

<h4> Inheritance </h4>

* Inheritance is the most important aspect of object-oriented programming, which simulates the real-world concept of inheritance. It specifies that the child object acquires all the properties and behaviors of the parent object.

* By using inheritance, we can create a class which uses all the properties and behavior of another class. The new class is known as a derived class or child class, and the one whose properties are acquired is known as a base class or parent class.

<h4> Polymorphism </h4>

* Polymorphism contains two words "poly" and "morphs". Poly means many, and morph means shape. By polymorphism, we understand that one task can be performed in different ways. For example - you have a class animal, and all animals speak. But they speak differently. Here, the "speak" behavior is polymorphic in a sense and depends on the animal. So, the abstract "animal" concept does not actually "speak", but specific animals (like dogs and cats) have a concrete implementation of the action "speak".


<h4> Encapsulation</h4>

* Encapsulation is also an essential aspect of object-oriented programming. It is used to restrict access to methods and variables. In encapsulation, code and data are wrapped together within a single unit from being modified by accident.

<h4>Data Abstraction </h4>

* Data abstraction and encapsulation both are often used as synonyms. Both are nearly synonyms because data abstraction is achieved through encapsulation.

* Abstraction is used to hide internal details and show only functionalities. Abstracting something means to give names to things so that the name captures the core of what a function or a whole program does.